src/axom/spin/internal/linear_bvh/build_radix_tree.hpp

```diff
@@ -0,0 +1,685 @@
+template< typename ExecSpace, typename T>
+static void array_memset(T* array, const int32 size, const T val)
+{
+  using exec_policy = typename spin::execution_space< ExecSpace >::raja_exec;
+  RAJA::forall< exec_policy >(
+      RAJA::RangeSegment(0, size), AXOM_LAMBDA (int32 i)
+  {
+    array[ i ] = val;
+  } );
+}
+
+//------------------------------------------------------------------------------
+template < typename ExecSpace, typename FloatType, int NDIMS >
+void propagate_aabbs( RadixTree< FloatType, NDIMS >& data)
+{
+  const int inner_size = data.m_inner_size;
+  const int leaf_size = data.m_inner_size + 1;
+  SLIC_ASSERT( leaf_size == data.m_size );
+
+  // Pointers and vars are redeclared because I have a faint memory
+  // of a huge amount of pain and suffering due so cuda
+  // labda captures of pointers indide a struct. Bad memories
+  // of random segfaults ........ be warned
+  const int32 *lchildren_ptr = data.m_left_children;
+  const int32 *rchildren_ptr = data.m_right_children;
+  const int32 *parent_ptr = data.m_parents;
+  const AABB< FloatType,NDIMS >  *leaf_aabb_ptr = data.m_leaf_aabbs;
+
+  AABB<FloatType,NDIMS>  *inner_aabb_ptr = data.m_inner_aabbs;
+
+  int32* counters_ptr = axom::allocate<int32>(inner_size);
+
+  array_memset< ExecSpace >(counters_ptr, inner_size, 0);
+
+  using exec_policy   =
+      typename spin::execution_space< ExecSpace >::raja_exec;
+  using atomic_policy =
+      typename spin::execution_space< ExecSpace >::raja_atomic;
+
+  RAJA::forall< exec_policy >(
+      RAJA::RangeSegment(0,leaf_size), AXOM_LAMBDA(int32 i)
+  {
+    int32 current_node = parent_ptr[inner_size + i];
+
+    while( current_node != -1)
+    {
+      int32 old= RAJA::atomic::atomicAdd< atomic_policy >(&(counters_ptr[current_node]),1);
+
+      if(old == 0)
+      {
+        // first thread to get here kills itself
+        return;
+      }
+
+      int32 lchild = lchildren_ptr[current_node];
+      int32 rchild = rchildren_ptr[current_node];
+
+      // gather the aabbs
+      AABB< FloatType,NDIMS > aabb;
+      if(lchild >= inner_size)
+      {
+        aabb.include(leaf_aabb_ptr[lchild - inner_size]);
+      }
+      else
+      {
+        aabb.include(inner_aabb_ptr[lchild]);
+      }
+
+      if(rchild >= inner_size)
+      {
+        aabb.include(leaf_aabb_ptr[rchild - inner_size]);
+      }
+      else
+      {
+        aabb.include(inner_aabb_ptr[rchild]);
+      }
+
+      inner_aabb_ptr[current_node] = aabb;
+
+      current_node = parent_ptr[current_node];
+    }
+
+  } );
+
+  axom::deallocate(counters_ptr);
+}
+
+
+//------------------------------------------------------------------------------
+template < typename ExecSpace, typename FloatType, int NDIMS >
+void build_radix_tree( const FloatType* boxes,
+                       int size,
+                       AABB< FloatType, NDIMS >& bounds,
+                       RadixTree< FloatType, NDIMS >& radix_tree )
+{
+  radix_tree.allocate( size );
```

Comment:
`assert(size > 0)`

Reply:
I added SLIC_ASSERTs for sanity checking.